### PR TITLE
Adding schedule CRD and controller

### PR DESCRIPTION
**Pull request by sahare**
1) Using Velero schedule instead of Velero backup with user defined intervals: https://github.com/open-cluster-management/backlog/issues/15065
2) Should resolve most of the sonar smells regarding lines greater than 120 chars

Following items are outstanding and will be taken care of in separate PRs:

- Validating webhooks for schedule expression (https://github.com/open-cluster-management/backlog/issues/15699)
- Handling backups that need to be done only once
- Adding the rules related to this new CRD to the chart

### CODE REVIEW

**Comment by sdminonne** [comment]
@sahare as soon you think I can have a look at this, please let me know.

**Comment by sahare** [comment]
@sdminonne Please review, and I will try to fix the tests, not sure why they failed.

File: api/v1beta1/schedule_types.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	veleroapi "github.com/vmware-tanzu/velero/pkg/apis/velero/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
```
Comment (birsanv) [inline api/v1beta1/schedule_types.go:25]: Let's remove this comment

**Review by birsanv — APPROVED** [review]
lgtm

File: api/v1beta1/schedule_types.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// BackupScheduleStatus defines the observed state of BackupSchedule
+type BackupScheduleStatus struct {
+	// Phase shows the status for the backup operation
+	// +kubebuilder:validation:Optional
+	Phase StatusPhase `json:"phase"`
+	// Velero Schedule for backing up remote clusters
+	// +kubebuilder:validation:Optional
+	VeleroScheduleManagedClusters *veleroapi.Schedule `json:"veleroManagedClustersSchedule,omitempty"`
```
Comment (sdminonne) [inline api/v1beta1/schedule_types.go:46]: I would prefer to be consistent with variable name and json tags... Either
`VeleroScheduleManagedClusters *veleroapi.Schedule `json:"veleroScheduleManagedClusters,omitempty"`
or 
`VeleroManagedClustersSchedule *veleroapi.Schedule `json:"veleroManagedClustersSchedule,omitempty"`

File: api/v1beta1/schedule_types.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	// Velero Schedule for backing up other resources
+	// +kubebuilder:validation:Optional
+	VeleroScheduleResources *veleroapi.Schedule `json:"veleroResourcesSchedule,omitempty"`
```
Comment (sdminonne) [inline api/v1beta1/schedule_types.go:49]: same

File: api/v1beta1/schedule_types.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	// Velero Schedule for backing up credentials
+	// +kubebuilder:validation:Optional
+	VeleroScheduleCredentials *veleroapi.Schedule `json:"veleroCredentialsSchedule,omitempty"`
```
Comment (sdminonne) [inline api/v1beta1/schedule_types.go:52]: same

File: config/rbac/role.yaml [hunk 5]
```diff
@@ -32,6 +32,32 @@ rules:
...
+  - update
+- apiGroups:
+  - cluster.open-cluster-management.io
+  resources:
+  - backupschedules/status
+  verbs:
+  - get
+  - patch
+  - update
 - apiGroups:
```
Comment (sdminonne) [inline config/rbac/role.yaml:61]: We don't need to `get,list,watch`? 

Here we're missing the tags I've just added in the fix for RBAC... Same.
Add `kubebuilder` tags for rabc for `managedclusters.cluster.open-cluster-management.io`

File: controllers/backup.go [hunk 6]
```diff
@@ -49,27 +49,30 @@ var (
...
-		}
+	// check if current backup resource exists
+	veleroIdentity := types.NamespacedName{
+		Namespace: backup.Spec.VeleroConfig.Namespace,
+		Name:      backup.Status.CurrentBackup,
+	}
+	// get the velero CR using the veleroIdentity
+	veleroBackup := &veleroapi.Backup{}
+	err := r.Get(ctx, veleroIdentity, veleroBackup)
+	if err != nil {
```
Comment (sdminonne) [inline controllers/backup.go:72]: you can create new one if the velero backup doesn't exists... But you should check the `err` type. If it's a `NotFound` but you may have other errors (for example `APISever` cannot answer`... Unsure If I understand the logic correctly but at least the comment is misleading

File: controllers/schedule_controller.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+// BackupScheduleReconciler reconciles a BackupSchedule object
+type BackupScheduleReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=backupschedules,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=backupschedules/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=cluster.open-cluster-management.io,resources=backupschedules/finalizers,verbs=update
+
```
Comment (sdminonne) [inline controllers/schedule_controller.go:68]: add other rbacs here and then `make manifests` to update role yaml file

File: controllers/schedule_controller.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.8.3/pkg/reconcile
+func (r *BackupScheduleReconciler) Reconcile(
+	ctx context.Context,
+	req ctrl.Request,
+) (ctrl.Result, error) {
+	scheduleLogger := log.FromContext(ctx)
+	backupSchedule := &v1beta1.BackupSchedule{}
+
+	if err := r.Get(ctx, req.NamespacedName, backupSchedule); err != nil {
+		// we can get not-found errors on deleted requests
```
Comment (sdminonne) [inline controllers/schedule_controller.go:86]: yes, thanks remove this comment :)

File: controllers/schedule_controller.go [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+			"name", req.Name,
+		)
+		return ctrl.Result{}, err
+	}
+
+	if len(veleroScheduleList.Items) == 0 {
+		if err := r.initVeleroSchedules(ctx, backupSchedule, r.Client); err != nil {
+			return ctrl.Result{}, err
+		}
+	} else {
```
Comment (sdminonne) [inline controllers/schedule_controller.go:106]: why the `else`? `golang` doesn't like `if-then-else` and in this case it's useless :)

**Review by sdminonne — APPROVED** [review]
There're things that I would change. `rbac`, some comments or logic and name and json tags. I would say that we should merge this PR and keep working on this.

**Comment by sdminonne** [comment]
@sahare done. Hope it makes sense :)

File: controllers/backup.go (hunk 6 above)
Comment (sahare) [reply controllers/backup.go:72]: This is just formatting change to reduce line length and remove useless else block, no changes in the logic here, so I would keep it as is in this PR :)

File: config/rbac/role.yaml [hunk 10]
```diff
@@ -58,6 +58,22 @@ rules:
...
+  - list
+  - watch
+- apiGroups:
+  - cluster.open-cluster-management.io
+  resources:
+  - managedclusters
+  verbs:
+  - get
+  - list
+  - watch
```
Comment (sdminonne) [inline config/rbac/role.yaml:76]: Thanks I missed this :)

File: controllers/schedule_controller.go [hunk 11]
```diff
@@ -99,14 +101,15 @@ func (r *BackupScheduleReconciler) Reconcile(
...
-		for _, veleroSchedule := range veleroScheduleList.Items {
-			updateScheduleStatus(ctx, &veleroSchedule, backupSchedule)
-		}
+	}
+	// update schedule status with latest velero schedules
+	for _, veleroSchedule := range veleroScheduleList.Items {
+		updateScheduleStatus(ctx, &veleroSchedule, backupSchedule)
 	}
 
 	err := r.Client.Status().Update(ctx, backupSchedule)
```
Comment (sdminonne) [inline controllers/schedule_controller.go:115]: Do we always need to update the status? Only if at line 112  the backupSchedule changed right? But we can add it later on...

**Review by sdminonne — APPROVED** [review]
Cool, thanks. This is mergeable.

File: controllers/schedule_controller.go (hunk 11 above)
Comment (sahare) [reply controllers/schedule_controller.go:115]: Sure, I'll compare to only update it if it's changed in my next PR :) Thanks!

**Pull request merged**